dask/array/core.py

```diff
@@ -3493,6 +3498,72 @@ def _vindex(x, *indexes):
     return result_1d.reshape(broadcast_shape + result_1d.shape[1:])
 
 
+def _vindex_dask_arrays(x, *indices):
+    """Point wise indexing with Dask Arrays"""
+
+    from .creation import arange
+
+    x = asarray(x)
+
+    oth_inds = []
+    arr_inds = []
+    out_axes = []
+    for i, ind in enumerate(indices):
+        if isinstance(ind, Array):
+            oth_inds.append(slice(None))
+            arr_inds.append(asarray(ind))
+            if 0 not in out_axes:
+                out_axes.append(0)
+        elif isinstance(ind, Integral):
+            oth_inds.append(ind)
+        else:
+            oth_inds.append(ind)
+            arr_inds.append(slice(None))
+            out_axes.append(len(arr_inds))
+    oth_inds = tuple(oth_inds)
+    arr_inds = tuple(arr_inds)
+    out_axes = tuple(out_axes)
+
+    if not all(o == slice(None) for o in oth_inds):
+        x = x.vindex[oth_inds]
```

Comment:
OK, so we do this in two passes, first with `other_indices` and then with the `array_indices`?

Reply:
Right. Sorry should have been explicit about this before.

We do the first pass to deal with everything that is static (e.g. known single indexes, slices, lists, NumPy arrays). Then we do a second pass with everything that is dynamic (i.e. Dask Arrays of known and unknown length).

Assumed there are use cases (xarray?) where being able to explicitly select portions of the graph to keep are important. If not, we could adjust our `atop` call to handle these as well.

Reply:
Yes, there are probably use cases where the ability to identify the right chunks ahead of time would make a big difference for efficiency.

Reply:
Yep, that would not surprise me. :)

Might do some light refactoring in here to make the static and dynamic passes more obvious.

Reply:
Added a comment before this to make the first pass obvious. Also explained in the docstring now.